### PR TITLE
refactor: combined capacity updates and decoding for all physical decoders

### PR DESCRIPTION
**Pull request by raunaks13**
- Combines `update_capacity` and `decode_into` into a single function `decode` for all physical (primitive) decoders (`BasicDecoder`, `BinaryPageDecoder`, `ValuePageDecoder`, `FixedListDecoder`, `BitmapDecoder`)
- As a result some of the decoders don't require explicit capacity allocation at a certain level. Should enable decoders to have more flexibility in their decode pipeline (required for dictionary encoding)
- Buffers built up from leaves of the encoding tree and combined instead of being recursively passed down (like earlier)
- Rename PhysicalPageDecoder to PrimitivePageDecoder

### CODE REVIEW

File: rust/lance-encoding/src/decoder.rs [hunk 1]
```diff
@@ -1165,13 +1165,13 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
-        buffers: &mut [(u64, bool)],
-        all_null: &mut bool,
-    );
+    // fn update_capacity(
+    //     &self,
+    //     rows_to_skip: u32,
+    //     num_rows: u32,
+    //     buffers: &mut [(u64, bool)],
+    //     all_null: &mut bool,
+    // );
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1174]:
```suggestion
```

File: rust/lance-encoding/src/decoder.rs [hunk 2]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
     /// * `rows_to_skip` - how many rows to skip (within the page) before decoding
     /// * `num_rows` - how many rows to decode
     /// * `dest_buffers` - the output buffers to decode into
+    // fn decode_into(
+    //     &self,
+    //     rows_to_skip: u32,
+    //     num_rows: u32,
+    //     dest_buffers: &mut [BytesMut],
+    // ) -> Result<()>;
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1190]:
```suggestion
```

Also, remove all the old rustdoc comments for these methods.

File: rust/lance-encoding/src/decoder.rs [hunk 3]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+
+    /// Combined function to update the capacity required to represent the requested data
+    /// and decode the data into the requested buffers.
+    ///
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1194]: No need to describe it as a "combined function".  The fact that we are combining two functions into one will hopefully be lost to history.  The first sentence of a method doc should be a brief summary of what the method does.

```suggestion
    /// Decode data into buffers
    ///
    /// This may be a simple zero-copy from a disk buffer or could involve complex decoding
    /// such as decompressing from some compressed representation.
```

File: rust/lance-encoding/src/decoder.rs [hunk 4]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+    /// Capacity is stored as a tuple of (num_bytes: u64, is_needed: bool).  The `is_needed`
+    /// portion only needs to be updated if the encoding has some concept of an "optional"
+    /// buffer.
+    ///
+    /// The decoder should look at `rows_to_skip` and `num_rows` and then calculate how
+    /// many bytes of data are needed.  It should then update the first part of the tuple.
+    ///
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1201]:
```suggestion
```

File: rust/lance-encoding/src/decoder.rs [hunk 5]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+    /// Note: Most encodings deal with a single buffer.  They may have multiple input buffers
+    /// but they only have a single output buffer.  The current exception to this rule is the
+    /// `basic` encoding which has an output "validity" buffer and an output "values" buffers.
+    /// We may find there are other such exceptions.
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1205]:
```suggestion
    /// Encodings can have any number of input or output buffers.  For example, a dictionary
    /// decoding will convert multiple two buffers (indices + dictionary) into a single buffer
    ///
    /// Binary decodings have two output buffers (one for values, one for offsets)
    ///
    /// Other decodings could even expand the # of output buffers.  For example, we could decode
    /// fixed size strings into variable length strings going from one input buffer to multiple output
    /// buffers.
    ///
    /// Each Arrow data type typically has a fixed structure of buffers and the encoding chain will
    /// generally end at one of these structures.  However, intermediate structures may exist which
    /// do not correspond to any Arrow type at all.  For example, a bitpacking encoding will deal
    /// with buffers that have bits-per-value that is not a multiple of 8.
    ///
    /// The `primitive_array_from_buffers` method has an expected buffer layout for each arrow
    /// type (order matters) and encodings that aim to decode into arrow types should respect
    /// this layout.
```

The old comment was getting outdated

File: rust/lance-encoding/src/decoder.rs [hunk 6]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+    /// # Arguments
+    ///
+    /// * `rows_to_skip` - how many rows to skip (within the page) before decoding
+    /// * `num_rows` - how many rows to decode
+    /// * `capacities` - A mutable slice of "capacities" (as described above), one per buffer
+    /// * `dest_buffers` - the output buffers to decode into
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1211]:
```suggestion
```

File: rust/lance-encoding/src/decoder.rs [hunk 7]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+    /// * `all_null` - A mutable bool, set to true if a decoder determines all values are null
     fn decode_into(
         &self,
         rows_to_skip: u32,
         num_rows: u32,
-        dest_buffers: &mut [BytesMut],
-    ) -> Result<()>;
+        // capacities: &mut [(u64, bool)],
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1217]:
```suggestion
```

File: rust/lance-encoding/src/decoder.rs [hunk 8]
```diff
@@ -1182,12 +1182,41 @@ pub trait PhysicalPageDecoder: Send + Sync {
...
+    /// `basic` encoding which has an output "validity" buffer and an output "values" buffers.
+    /// We may find there are other such exceptions.
+    /// # Arguments
+    ///
+    /// * `rows_to_skip` - how many rows to skip (within the page) before decoding
+    /// * `num_rows` - how many rows to decode
+    /// * `capacities` - A mutable slice of "capacities" (as described above), one per buffer
+    /// * `dest_buffers` - the output buffers to decode into
+    /// * `all_null` - A mutable bool, set to true if a decoder determines all values are null
     fn decode_into(
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1213]: Let's rename this method to `decode`.  We are no longer "decoding into" any existing data.

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 9]
```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
...
-        buffers[0].1 = matches!(self.mode, DataNullStatus::Some(_));
-        if let Some(values) = self.mode.values_decoder() {
-            values.update_capacity(rows_to_skip, num_rows, &mut buffers[1..], all_null);
-        } else {
-            *all_null = true;
-        }
-    }
+    ) -> Result<Vec<BytesMut>> {
+        if let Some(_values) = self.mode.values_decoder() {
+            match &self.mode {
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:180]: I think you should match on mode first.  Then you can just unwrap `self.mode.values_decoder()` when you need it.  The end result should be more readable.

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 10]
```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
...
+                    capacities[0].0 = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
+                    // The validity buffer is only required if we have some nulls
+                    capacities[0].1 = matches!(self.mode, DataNullStatus::Some(_));
+
+                    let mut dest_buffers = create_buffers_from_capacities(capacities);
+
+                    dest_buffers[0].fill(0);
+
+                    Ok(dest_buffers)
+                }
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:209]: This is unreachable.  If we are in this mode then `self.mode.values_decoder` is `None` in which case we will not be in this if branch.  Is this code important?  If so, it probably belongs in the `else` branch (although the whole top-level if/else branch should just go away per previous comment)

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 11]
```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
...
+                DataNullStatus::None(values) => {
+                    let mut capacities = vec![(0, false); self.num_buffers() as usize];
+
+                    // No need to look at the validity decoder to know the dest buffer size since it is boolean
+                    capacities[0].0 = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
+                    // The validity buffer is only required if we have some nulls
+                    capacities[0].1 = matches!(self.mode, DataNullStatus::Some(_));
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:216]: How can this be true if we are in `DataNullStatus::None`?

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 12]
```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
...
+                    let mut dest_buffers = create_buffers_from_capacities(capacities);
+
+                    dest_buffers[0].fill(1);
+                    let values_bytesmut = values.decode_into(rows_to_skip, num_rows, all_null)?; // buffer 1 onwards
+
+                    for (i, value) in values_bytesmut.into_iter().enumerate() {
+                        dest_buffers[i + 1] = value;
+                    }
+
+                    Ok(dest_buffers)
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:227]: This is overly complicated.  It should just be...

```
let mut buffers = vec![BytesMut::new()];
buffers.extend(values.decode_into(rows_to_skip, num_rows, all_null)?;
Ok(buffers)
```

File: rust/lance-encoding/src/encodings/physical/binary.rs [hunk 13]
```diff
@@ -245,18 +226,21 @@ impl PhysicalPageDecoder for BinaryPageDecoder {
...
-        dest_buffers[0].extend_from_slice(byte_slice);
+        // dest_buffers[0].extend_from_slice(byte_slice);
 
         // Copy decoded bytes into dest_buffers[1..]
         // Currently an empty null buffer is the first one
         // The actual bytes are in the second buffer
         // Including the indices this results in 3 buffers in total
-        self.bytes_decoder
-            .decode_into(bytes_to_skip, num_bytes, &mut dest_buffers[1..])?;
+        // dest_buffers.append(&mut bytes_buffers);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/binary.rs:244]: Clean up these commented code blocks

File: rust/lance-encoding/src/encodings/physical/bitmap.rs [hunk 14]
```diff
@@ -92,24 +93,20 @@ struct BitmapDecoder {
...
-        dest_buffers: &mut [BytesMut],
-    ) -> Result<()> {
+        _all_null: &mut bool,
+    ) -> Result<Vec<BytesMut>> {
+        let mut capacities = vec![(0, false); self.num_buffers() as usize];
+
+        capacities[0].0 = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
+        capacities[0].1 = true;
+
+        let mut dest_buffers = create_buffers_from_capacities(capacities);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/bitmap.rs:108]: I think we can ditch the `create_buffers_from_capacities` style.  With this new approach we don't really need to create an arbitrary number of buffers.  We pretty much always know how many we need.

```suggestion
        let num_bytes = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
        let buffer = BytesMut::with_capacity(num_bytes);
```

File: rust/lance-encoding/src/encodings/physical/bitmap.rs [hunk 15]
```diff
@@ -92,24 +93,20 @@ struct BitmapDecoder {
...
+
         let mut rows_to_skip = rows_to_skip;
 
         let mut dest_builder = BooleanBufferBuilder::new(num_rows as usize);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/bitmap.rs:112]: Hmm...going forward I think we can actually make this zero-copy now.  But lets save that for a future PR :)

File: rust/lance-encoding/src/lib.rs [hunk 16]
```diff
@@ -10,7 +10,11 @@ use lance_core::Result;
 
 pub mod decoder;
 pub mod encoder;
-pub mod encodings;
+pub mod encodings {
+    pub mod logical;
+    pub mod physical;
+    pub mod utils;
+}
```
Comment (westonpace) [inline rust/lance-encoding/src/lib.rs:17]: Why this change?

**Review by westonpace — CHANGES_REQUESTED** [review]
This is a good idea but I think we can further simplify things.

File: rust/lance-encoding/src/lib.rs (hunk 16 above)
Comment (raunaks13) [reply rust/lance-encoding/src/lib.rs:17]: For the import from utils.rs
But I'll delete the method based on previous comments, so it won't be necessary any longer

Comment (westonpace) [reply rust/lance-encoding/src/lib.rs:17]: In the future you can just add `utils` to `rust/lance-encodings/src/encodings.rs`

File: rust/lance-encoding/src/decoder.rs [hunk 17]
```diff
@@ -1144,50 +1144,44 @@ impl BatchDecodeStream {
...
     ///
-    /// The decoder should look at `rows_to_skip` and `num_rows` and then calculate how
-    /// many bytes of data are needed.  It should then update the first part of the tuple.
-    ///
-    /// Note: Most encodings deal with a single buffer.  They may have multiple input buffers
-    /// but they only have a single output buffer.  The current exception to this rule is the
-    /// `basic` encoding which has an output "validity" buffer and an output "values" buffers.
-    /// We may find there are other such exceptions.
+    /// Encodings can have any number of input or output buffers.  For example, a dictionary
+    /// decoding will convert multiple two buffers (indices + dictionary) into a single buffer
```
Comment (westonpace) [inline rust/lance-encoding/src/decoder.rs:1158]:
```suggestion
    /// decoding will convert two buffers (indices + dictionary) into a single buffer
```

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 18]
```diff
@@ -167,51 +168,44 @@ struct BasicPageDecoder {
...
+
+                buffers.append(&mut values_bytesmut);
+                buffers
             }
             // Either dest_buffers[0] is empty, in which case these are no-ops, or one of the
             // other pages needed the buffer, in which case we need to fill our section
             DataNullStatus::All => {
-                dest_buffers[0].fill(0);
+                let mut buffers = vec![BytesMut::default()];
+                buffers[0].fill(0);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:191]:
```suggestion
```

Why is this call needed?  You've declared an empty buffer so this should be a no-op.

In the old code we had a special path for "I'm all null but not everyone else is and so I'd better fill in the bits with zero".  This path no longer exists because each page does its own thing now.

File: rust/lance-encoding/src/encodings/physical/basic.rs [hunk 19]
```diff
@@ -167,51 +168,44 @@ struct BasicPageDecoder {
...
+                *all_null = true;
+                buffers
             }
             DataNullStatus::None(values) => {
+                let mut dest_buffers = vec![BytesMut::with_capacity(arrow_buffer::bit_util::ceil(
+                    num_rows as usize,
+                    8,
+                ))];
                 dest_buffers[0].fill(1);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/basic.rs:200]: Hmm...in the future we should be able to optimize this away and declare an empty buffer.  Does it work today?  I think I know one reason it might not work today but lets not worry about it yet if that doesn't work.

File: rust/lance-encoding/src/encodings/physical/bitmap.rs [hunk 20]
```diff
@@ -189,8 +181,8 @@ mod tests {
                 },
             ],
         };
-        let mut dest = vec![BytesMut::with_capacity(1)];
-        let result = decoder.decode_into(5, 1, &mut dest);
+
+        let result = decoder.decode(5, 1, &mut false);
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/bitmap.rs:185]: I think we can now change this from a mutable input to a part of the output.  Let's fix in a future PR.

**Review by westonpace — APPROVED** [review]
Awesome, thanks for cleaning this up.  A few more tweaks and we are good to go.  There are a few minor things we can investigate for perf improvement in the future.

File: rust/lance-encoding/src/encodings/physical/basic.rs (hunk 19 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encodings/physical/basic.rs]: Works today, made the change
